File: src/providers/integraicp/state.ts
import { randomUUID } from "node:crypto";
import { getDb } from "../../shared/storage/db.js";
import type { FakeCertificate } from "./crypto.js";

export interface IntegraIcpCredential {
  id: string;
  channelId: string;
  subjectKey: string | null;
  subjectType: string | null;
  codeChallenge: string;
  callbackUri: string;
  certificate: FakeCertificate;
  createdAt: string;
}

interface CredentialRow {
  id: string;
  channel_id: string;
  subject_key: string | null;
  subject_type: string | null;
  code_challenge: string;
  callback_uri: string;
  certificate_json: string;
  created_at: string;
}

function toCredential(row: CredentialRow): IntegraIcpCredential {
  return {
    id: row.id,
    channelId: row.channel_id,
    subjectKey: row.subject_key,
    subjectType: row.subject_type,
    codeChallenge: row.code_challenge,
    callbackUri: row.callback_uri,
    certificate: JSON.parse(row.certificate_json),
    createdAt: row.created_at,
  };
}

export function createCredential(input: {
  channelId: string;
  subjectKey?: string;
  subjectType?: string;
  codeChallenge: string;
  callbackUri: string;
  certificate: FakeCertificate;
}): IntegraIcpCredential {
  const db = getDb();
  const id = randomUUID();
  db.prepare(
    `INSERT INTO integraicp_credentials (id, channel_id, subject_key, subject_type, code_challenge, callback_uri, certificate_json)
     VALUES (?, ?, ?, ?, ?, ?, ?)`,
  ).run(
    id,
    input.channelId,
    input.subjectKey ?? null,
    input.subjectType ?? null,
    input.codeChallenge,
    input.callbackUri,
    JSON.stringify(input.certificate),
  );
  const row = db.prepare("SELECT * FROM integraicp_credentials WHERE id = ?").get(id) as CredentialRow;
  return toCredential(row);
}

export function getCredential(id: string): IntegraIcpCredential | undefined {
  const row = getDb().prepare("SELECT * FROM integraicp_credentials WHERE id = ?").get(id) as
    | CredentialRow
    | undefined;
  return row ? toCredential(row) : undefined;
}

File: internal/providers/integraicp/state.go
package integraicp

import (
	"database/sql"
	"encoding/json"
	"errors"

	"github.com/google/uuid"
)

type Credential struct {
	ID            string          `json:"id"`
	ChannelID     string          `json:"channelId"`
	SubjectKey    *string         `json:"subjectKey"`
	SubjectType   *string         `json:"subjectType"`
	CodeChallenge string          `json:"codeChallenge"`
	CallbackURI   string          `json:"callbackUri"`
	Certificate   FakeCertificate `json:"certificate"`
	CreatedAt     string          `json:"createdAt"`
}

type Store struct {
	db *sql.DB
}

func NewStore(db *sql.DB) *Store {
	return &Store{db: db}
}

func (s *Store) scanCredential(row interface {
	Scan(dest ...any) error
}) (*Credential, error) {
	var (
		c        Credential
		certJSON string
	)
	if err := row.Scan(&c.ID, &c.ChannelID, &c.SubjectKey, &c.SubjectType, &c.CodeChallenge, &c.CallbackURI, &certJSON, &c.CreatedAt); err != nil {
		return nil, err
	}
	if err := json.Unmarshal([]byte(certJSON), &c.Certificate); err != nil {
		return nil, err
	}
	return &c, nil
}

type CreateCredentialInput struct {
	ChannelID     string
	SubjectKey    *string
	SubjectType   *string
	CodeChallenge string
	CallbackURI   string
	Certificate   FakeCertificate
}

func (s *Store) CreateCredential(input CreateCredentialInput) (*Credential, error) {
	id := uuid.NewString()
	certJSON, err := json.Marshal(input.Certificate)
	if err != nil {
		return nil, err
	}
	_, err = s.db.Exec(
		`INSERT INTO integraicp_credentials (id, channel_id, subject_key, subject_type, code_challenge, callback_uri, certificate_json)
		 VALUES (?, ?, ?, ?, ?, ?, ?)`,
		id, input.ChannelID, input.SubjectKey, input.SubjectType, input.CodeChallenge, input.CallbackURI, string(certJSON),
	)
	if err != nil {
		return nil, err
	}
	row := s.db.QueryRow(
		`SELECT id, channel_id, subject_key, subject_type, code_challenge, callback_uri, certificate_json, created_at
		 FROM integraicp_credentials WHERE id = ?`, id,
	)
	return s.scanCredential(row)
}

func (s *Store) ListCredentials() ([]Credential, error) {
	rows, err := s.db.Query(
		`SELECT id, channel_id, subject_key, subject_type, code_challenge, callback_uri, certificate_json, created_at
		 FROM integraicp_credentials ORDER BY created_at DESC`,
	)
	if err != nil {
		return nil, err
	}
	defer func() { _ = rows.Close() }()

	out := []Credential{}
	for rows.Next() {
		c, err := s.scanCredential(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, *c)
	}
	return out, rows.Err()
}

func (s *Store) GetCredential(id string) (*Credential, bool, error) {
	row := s.db.QueryRow(
		`SELECT id, channel_id, subject_key, subject_type, code_challenge, callback_uri, certificate_json, created_at
		 FROM integraicp_credentials WHERE id = ?`, id,
	)
	c, err := s.scanCredential(row)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, false, nil
	}
	if err != nil {
		return nil, false, err
	}
	return c, true, nil
}
